espresso-tests/migration-test.bash

```diff
@@ -0,0 +1,146 @@
+
+echo "Deployed new OspMigrationAction at $OSP_MIGRATION_ACTION"
+
+# use cast to call the upgradeExecutor to execute the l1 upgrade actions.
+
+cast send $UPGRADE_EXECUTOR "execute(address, bytes)" $OSP_MIGRATION_ACTION $(cast calldata "perform()") --rpc-url $RPC_URL --private-key $PRIVATE_KEY
+
+echo "Executed OspMigrationAction via UpgradeExecutor"
+
+NUM_CONFIRMED_NODES_BEFORE_UPGRADE=$(cast call --rpc-url $RPC_URL $ROLLUP_ADDRESS 'latestConfirmed()(uint256)')
+#shutdown nitro node
+docker stop nitro-testnode-sequencer-1
+
+#start nitro node in new docker container with espresso image (create a new script to do this from pieces of test-node.bash)
+cd $TESTNODE_DIR
+./espresso-tests/create-espresso-integrated-nitro-node.bash
+
+# Wait for RPC_URL to be available
+while ! curl -s $L2_RPC_URL > /dev/null; do
+  echo "Waiting for $L2_RPC_URL to be available..."
+  sleep 5
+done
+
+echo "Adding child chain upgrade executor as an L2 chain owner"
+cast send 0x0000000000000000000000000000000000000070 'addChainOwner(address)' $CHILD_CHAIN_UPGRADE_EXECUTOR_ADDRESS --rpc-url $L2_RPC_URL --private-key $PRIVATE_KEY
+
+echo "Deploying ArbOS action"
+
+#forge script to deploy the Espresso ArbOS upgrade action.
+cd $ORBIT_ACTIONS_DIR
+forge script --chain $L2_CHAIN_NAME contracts/child-chain/arbos-upgrade/DeployArbOSUpgradeAction.s.sol:DeployArbOSUpgradeAction  --rpc-url $L2_RPC_URL --broadcast -vvvv
+ARBOS_UPGRADE_ACTION=$(cat broadcast/DeployArbOSUpgradeAction.s.sol/412346/run-latest.json | jq -r '.transactions[0].contractAddress')
+echo "Deployed ArbOSUpgradeAction at $ARBOS_UPGRADE_ACTION"
+
+# TODO: figure out why this upgrade does not do anything
+cast send $CHILD_CHAIN_UPGRADE_EXECUTOR_ADDRESS "execute(address, bytes)" $ARBOS_UPGRADE_ACTION $(cast calldata "perform()") --rpc-url $L2_RPC_URL --private-key $PRIVATE_KEY
+
+ARBOS_VERSION_BEFORE_UPGRADE=$(cast call "0x0000000000000000000000000000000000000064" "arbOSVersion()(uint64)" --rpc-url $L2_RPC_URL)
+
+# Check the upgrade happened
+
+ARBOS_VERSION_AFTER_UPGRADE=$(cast call "0x0000000000000000000000000000000000000064" "arbOSVersion()(uint64)" --rpc-url $L2_RPC_URL)
+
+while [ $ARBOS_VERSION_BEFORE_UPGRADE == $ARBOS_VERSION_AFTER_UPGRADE ]
+do
+  sleep 5
+  ARBOS_VERSION_AFTER_UPGRADE=$(cast call "0x0000000000000000000000000000000000000064" "arbOSVersion()(uint64)" --rpc-url $L2_RPC_URL)
+done
+
+# The arbsys precompile is returning 55 for the arbos version which indicates that the value internally is 0
+# We are upgrading the ArbOS version to 35 so the expect the return value to be 55 + 35 = 90
+if [ $ARBOS_VERSION_AFTER_UPGRADE != "90" ]; then
+  fail "ArbOS version not updated: Expected 90, Actual $ARBOS_VERSION_AFTER_UPGRADE"
+fi
+
+#test for new OSP address
+CHALLENGE_MANAGER_OSP_ADDRESS=$(cast call $CHALLENGE_MANAGER_ADDRESS "osp()(address)" --rpc-url $RPC_URL)
+if [ $NEW_OSP_ENTRY != $CHALLENGE_MANAGER_OSP_ADDRESS ]; then
```

Comment:
This currently fails because the address from the broadcast artifact is not checksummed (it's all lowercase) whereas the address we get from cast is checksummed.

Reply:
Can use `cast to-checksum` E. g. 

```
cat orbit-actions/broadcast/DeployEspressoOsp.s.sol/1337/run-latest.json | jq -r '.transactions[4].contractAddress' | cast to-checksum
```

Reply:
Do we need to do this for the other addresses that we grab from the deployment files?

Reply:
We only need to do it if we compare addresses as strings. Alternatively we can convert everything to lowercase.